File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    answer = 0
    
    def check_one_char(word1, word2):
        check = 0
        for i in range(len(word1)):
            if word1[i] != word2[i]:
                check += 1
        if check == 1:
            return True
        else:
            return False
        
    def dfs(word, cnt, visited):
        nonlocal target, answer
        if word == target:
            if answer == 0:
                answer = cnt
            elif answer > cnt:
                answer = cnt
            return
        elif answer != 0 and cnt > answer:
            return
        
        for i in range(len(words)):
            if check_one_char(word, words[i]) and not visited[i]:
                visited[i] = True
                dfs(words[i], cnt + 1, visited)
                visited[i] = False
    
    visited = [False] * len(words)
    dfs(begin, 0, visited)
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    int answer = 0;
    
    public Boolean oneDifferentWord(String word1, String word2) {
        int check = 0;
        for(int i = 0; i < word1.length(); i++) {
            if (word1.charAt(i) != word2.charAt(i)) {
                check += 1;
            }
        }
        if (check == 1) return true;
        else return false;
    }
    
    public void dfs(String word, String target, String[] words, int count, List<Boolean> visited) {
        if (word.equals(target)) {
            if (answer == 0) answer = count;
            else if (answer > count) answer = count;
            return;
        } else if (answer != 0 && count > answer) return;
        
        for (int i = 0; i < words.length; i++) {
            if (oneDifferentWord(words[i], word) && !visited.get(i)) {
                visited.set(i, true);
                dfs(words[i], target, words, count + 1, visited);
                visited.set(i, false);
            }
        }
    }
    public int solution(String begin, String target, String[] words) {
        List<Boolean> visited = new ArrayList<>();
        for (int i = 0; i < words.length; i++) {
            visited.add(false);
        }
        
        dfs(begin, target, words, 0, visited);
        return answer;
    }
}